src/components/TodoApp/TodoApp.tsx

```diff
@@ -0,0 +1,270 @@
+    settextInFormToCreateTodo(event.target.value);
+  };
+
+  useEffect(() => {
+    refInputCreatTodo.current!.focus();
+  });
+
+  useEffect(() => {
+    if (currentFilterCriteria !== ACTIONSINFILTERCONTEXT.ALL) {
+      setShowFooterOrNot(true);
+
+      return;
+    }
+
+    if (filteredTodos.length < 1) {
+      setShowFooterOrNot(false);
+
+      return;
+    }
+
+    setShowFooterOrNot(true);
+  }, [filteredTodos, currentFilterCriteria]);
+
+  const handleSubmitToCreateTodo = (
+    event: React.FormEvent<HTMLFormElement>,
+  ) => {
+    event.preventDefault();
+    refInputCreatTodo.current?.focus();
+    const whitNoWhitSpace = textInFormToCreateTodo.trim();
+
+    if (whitNoWhitSpace !== '') {
+      refInputCreatTodo.current!.disabled = true;
+
+      refreshTodos(prev => {
+        prev.push({
+          title: textInFormToCreateTodo,
+          id: 0,
+          userId: 0,
+          completed: false,
+          isAwaitServer: true,
+        });
+
+        return [...prev];
+      });
+
+      createTodo({ title: whitNoWhitSpace, userId: USER_ID, completed: false })
+        .then(res => {
+          refreshTodos(prev => {
+            const newState = [...prev];
+
+            newState[prev.length - 1] = { ...res, isAwaitServer: false };
+
+            return newState;
+          });
+          settextInFormToCreateTodo('');
+          refInputCreatTodo.current!.disabled = false;
+        })
+        .catch(() => {
+          putErrorWarning('Unable to add a todo');
+          refInputCreatTodo.current!.disabled = false;
+
+          todosFromServer.splice(
+            todosFromServer.findIndex(
+              e => e.id === todosFromServer[todosFromServer.length - 1].id,
+            ),
+            1,
+          );
+          refreshTodos([...todosFromServer]);
+          settextInFormToCreateTodo(prev => prev);
+        });
+    } else {
+      putErrorWarning('Title should not be empty');
+    }
+  };
+
+  const handleClearAllCompletedTodosClick = () => {
+    refreshTodos(prev => {
+      for (const elm of prev) {
+        if (elm.completed) {
+          elm.isAwaitServer = true;
+        }
+      }
+
+      return [...prev];
+    });
+
+    const todosToDelete = todosFromServer.reduce((acc, e) => {
+      if (e.completed) {
+        acc.push(e.id);
+      }
+
+      return acc;
+    }, [] as number[]);
+
+    const promises = todosToDelete.map(e => deleteTodo(e));
+    let realyDel: number[] = [];
+
+    Promise.allSettled(promises)
+      .then(res => {
+        let counter = 0;
+
+        if (res.some(e => e.status === 'rejected')) {
+          putErrorWarning('Unable to delete a todo');
+        }
+
+        realyDel = res.reduce((acc, e) => {
+          if (e.status === 'fulfilled') {
+            acc.push(todosToDelete[counter]);
+          }
+
+          counter++;
+
+          return acc;
+        }, [] as number[]);
+      })
+      .finally(() => {
+        realyDel.forEach(e => {
+          todosFromServer.splice(
```

Comment:
You are altering a external variable, which isn't good.
And this seems just like an overcomplicated _.filter_ implementation...

Reply:
I'll think about that and improve it, make sense.